### DISCUSSION TITLE
Ways to clear memory.

### DISCUSSION THREAD

**Discussion by maxfrei750** [mitsuba-renderer/mitsuba3#513] [Q&A] [answered]
I'm currently fighting with a high memory consumption when using the `path` integrator together with transparent materials and the `wrap_ad` decorator (from `drjit` to `torch`).

Using the following code, called between iterations, I was able to significantly reduce the problem:
```python
del image
gc.collect()
gc.collect()

torch.cuda.empty_cache()
```

With image being a previously rendered image. While looking for solutions, I found the `clean_up()` method in the code base:
https://github.com/mitsuba-renderer/mitsuba3/blob/152352f87b5baea985511b2a80d9f91c3c945a90/src/conftest.py#L45-L64

Now I'm wondering if some of these methods might help to further alleviate my problem. Unfortunately, I couldn't find a documentation for them. Therefore, I'd be grateful, if somebody could shed some light on them and assess if they are suitable to reduce memory load _while still being inside an optimization loop_, i.e. not mess with the optimization.

Thanks!

**Comment by njroussel** [comment, marked as answer]
Hi

The `path` integrator has a significant memory/efficiency cost to it when used in AD settings. I'd strongly recommend using `prb` instead, take a look at figure 6 of the [Path Replay Backpropagation paper](http://rgl.epfl.ch/publications/Vicini2021PathReplay). 
The `wrap_ad` function has some caveats to it but they rather have to do with runtime than memory consumption usually. 

I can't find the thread about it anymore, but typically `torch` also likes to reserve a huge chuck of your VRAM. So that might also be an avenue to look into.

The `clean_up()` method you found is very "strict". Because it's used between unit tests, it practically flushes the entire JIT's state which is rarely something you want. Anyway, this should not be a significant amount of memory and manually calling `del variable;gc.collect();gc.collect()` is the intended way to force variables to be deleted.

**Reply by maxfrei750** [reply]
Thanks for elaborating! I'd indeed prefer to use `prb`, but I'm facing a problem with exploding gradients (or something else that causes my parameters to sharply explode/drop), when I use it to optimize transparency, which is sort of expected, if I understand the [documentation](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_integrators.html#path-replay-backpropagation-prb) correctly:

> the properties of ideal specular objects (e.g., the IOR of a glass vase) cannot be optimized.

I tried to experiment with roughness and the `prb_reparam` integrator. But I'm not sure, if I understood the implications of the remark in the docs correctly.

**Reply by njroussel** [reply]
You might want to take a look at the discussion over [here](https://github.com/mitsuba-renderer/mitsuba3/discussions/501#discussioncomment-4843432) or read [this paper](http://rgl.epfl.ch/publications/Zeltner2021MonteCarlo). To understand this notion of attached/detached sampling.

**Reply by njroussel** [reply]
Oh I hadn't yet seen your comment in this thread: https://github.com/mitsuba-renderer/mitsuba3/discussions/379
There's a lot going on in this context, I can only recommend reading a bit more of the literature that has been linked in the documentation and these discussions.

**Reply by maxfrei750** [reply]
Will do. Thanks for pointing me into the right direction.
